File: js/290 Word Pattern/WordPattern.js

/**
 * Given a pattern and a string str, find if str follows the same pattern.
 * Here follow means a full match, such that there is a bijection between a letter in pattern and a non-empty word in str.
 */
 /**
  * @param {string} pattern
  * @param {string} str
  * @return {boolean}
  */
var wordPattern = function(pattern, str) {
	 const data = str.split(" ");
	 if (pattern.length != data.length)
		 return false;
	 let map = {};
	 for (let i = 0; i < pattern.length; i++) {
		 const key = pattern.charCodeAt(i), s = map[key], data_i = data[i];
		 if (s == null) {
			 for (let k in map)
				 if (map[k] == data_i)
					 return false;
			 map[key] = data_i;
		 } else if (s != data_i) {
			 return false;
		 }
	 }
	 return true;
}

File: go/290 Word Pattern/WordPattern.go

/*
 * Given a pattern and a string str, find if str follows the same pattern.
 * Here follow means a full match, such that there is a bijection between a letter in pattern and a non-empty word in str.
 */
import "strings"
func wordPattern(pattern string, str string) bool {
    data := strings.Split(str, " ")
    if len(pattern) != len(data) {return false}
    MAP := make(map[byte] string)
    for i, n := 0, len(pattern); i < n; i++ {
        key := pattern[i]
        s, ok := MAP[key]
        if !ok {
            data_i := data[i]
            for _, v := range MAP {
                if v == data_i { return false }
            }
            MAP[key] = data_i
        } else if s != data[i] {
            return false
        }
    }
    return true
}
